File: 20-ValidParentheses/20-ValidParentheses.cpp
// Last updated: 30/07/2026, 10:08:56
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;

        for (char c : s) {
            if (c == '(' || c == '[' || c == '{') {
                st.push(c);
            } else {
                if (st.empty()) return false;

                if ((c == ')' && st.top() != '(') ||
                    (c == ']' && st.top() != '[') ||
                    (c == '}' && st.top() != '{'))
                    return false;

                st.pop();
            }
        }

        return st.empty();
    }
};

File: 20-ValidParentheses/20-ValidParentheses.js
// Last updated: 30/07/2026, 10:09:01
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    const stack = [];
    const map = {
        ')': '(',
        ']': '[',
        '}': '{'
    };

    for (let char of s) {
        if (map[char]) {
            if (stack.pop() !== map[char]) {
                return false;
            }
        } else {
            stack.push(char);
        }
    }

    return stack.length === 0;
};

console.log(isValid("()"));     
console.log(isValid("()[]{}"));  
console.log(isValid("(]"));      
console.log(isValid("([])"));   
console.log(isValid("([)]"));    